### PR TITLE
Mrg segment

### PR DESCRIPTION
**Pull request by xpander53**

### CODE REVIEW

File: src/main/resources/config.properties [hunk 1]
```diff
@@ -1,4 +1,4 @@
 base.path.resource=
-supported.hl7.messages=ADT_A01, ORU_R01, PPR_PC1, VXU_V04, RDE_O11, OMP_O09, RDE_O25
+supported.hl7.messages=ADT_A01, ADT_A34, ADT_A40, ORU_R01, PPR_PC1, VXU_V04, RDE_O11, OMP_O09, RDE_O25
```
Comment (klwhaley) [inline src/main/resources/config.properties:2]: Does this pull request fully complete support for these two message types? if so then we can leave them, but if not we need to keep them locally.

Also, if they are fully supported, please update the readme to add them to the list

File: src/main/resources/hl7/message/ADT_A01.yml [hunk 2]
```diff
@@ -18,11 +18,12 @@ resources:
     - resourceName: Patient
       segment: PID
       resourcePath: resource/Patient
-      repeats: false
+      repeats: true
```
Comment (klwhaley) [inline src/main/resources/hl7/message/ADT_A01.yml:21]: do we need this ADT_A01? or only 34/40?

File: src/main/java/io/github/linuxforhealth/hl7/message/HL7MessageEngine.java [hunk 3]
```diff
@@ -93,17 +95,17 @@ public Bundle transform(final InputDataExtractor dataInput,
     Preconditions.checkArgument(resources != null, "resources cannot be null");
 
     HL7MessageData hl7DataInput = (HL7MessageData) dataInput;
-    Bundle bundle = initBundle();
+    var bundle = initBundle();
```
Comment (klwhaley) [inline src/main/java/io/github/linuxforhealth/hl7/message/HL7MessageEngine.java:98]: what is var?

File: src/main/resources/config.properties (hunk 1 above)
Comment (xpander53) [reply src/main/resources/config.properties:2]: I will remove them

File: src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+        // We should have link.other.reference to the PID (1st) patient
+        Base linkTwo = ResourceUtils.getValue(patientTwo, "link");
+        assertThat(ResourceUtils.getValueAsString(linkTwo, "type")).isEqualTo("Enumeration[replaced-by]");
+        Base patientTwoRef = ResourceUtils.getValue(linkTwo, "other");
+        assertThat(ResourceUtils.getValueAsString(patientTwoRef, "reference")).isEqualTo(patientOneId);
+
+    }
+
+    // Commented out until internal bug #684 is fixed.
```
[the remark this answers is not in the document]
Comment (xpander53) [reply src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java:154]: Great idea

File: src/main/resources/hl7/message/ADT_A01.yml (hunk 2 above)
Comment (xpander53) [reply src/main/resources/hl7/message/ADT_A01.yml:21]: I'm using ADT_A01 for now until the A40 bug is fixed.

File: src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (xpander53) [reply src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java]: Done

File: src/main/resources/hl7/resource/PatientMRG.yml [hunk 5]
```diff
@@ -0,0 +1,68 @@
+#
+# (C) Copyright IBM Corp. 2020, 2021
```
Comment (LisaWellman) [inline src/main/resources/hl7/resource/PatientMRG.yml:2]: remove 2020

File: src/main/resources/hl7/resource/PatientMRG.yml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+resourceType: Patient
+# Represents data that needs to be extracted for a Patient Resource in FHIR
```
Comment (LisaWellman) [inline src/main/resources/hl7/resource/PatientMRG.yml:7]: update comment to indicate this represents the patient in a MRG segment (outdated patient that should no longer be used)

File: src/main/resources/hl7/resource/PatientMRG.yml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+id:
+   type: STRING
+   valueOf: UUID.randomUUID()
+   expressionType: JEXL
+identifier_1:
+   required: true
+   valueOf: datatype/Identifier_SystemID
+   generateList: true
+   expressionType: resource
+   specs: MRG.1
```
Comment (LisaWellman) [inline src/main/resources/hl7/resource/PatientMRG.yml]: similar to above, should this be MRG1.1 ?

File: src/main/resources/hl7/resource/PatientMRG.yml [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+   vars:
+      assignerSystem: String, MRG.1
```
Comment (LisaWellman) [inline src/main/resources/hl7/resource/PatientMRG.yml]: Where is assignerSystem used?  I am not finding it
MRG1.4 is the assigning authority

File: src/main/resources/hl7/resource/PatientMRG.yml [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+# SPDX-License-Identifier: Apache-2.0
+#
+resourceType: Patient
+# Represents data that needs to be extracted for a Patient Resource in FHIR
+# reference: https://www.hl7.org/fhir/patient.html
+id:
+   type: STRING
+   valueOf: UUID.randomUUID()
+   expressionType: JEXL
+identifier_1:
```
Comment (LisaWellman) [inline src/main/resources/hl7/resource/PatientMRG.yml]: I am not understanding what this rule does (the second one makes more sense to me), maybe we can discuss.
Maybe this is the duplicate one we discussed in slack and it will be removed?
Any identifier from MRG should be marked as use:old

File: src/main/resources/hl7/resource/PatientMRG.yml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+      assignerSystem: String, MRG.1
+identifier_2:
+   condition: $valueIn NOT_NULL
+   valueOf: datatype/Identifier_var
+   generateList: true
+   expressionType: resource
+   vars:
+      valueIn: String, MRG.1
+   constants:
+      system: http://terminology.hl7.org/CodeSystem/v2-0203
```
Comment (LisaWellman) [inline src/main/resources/hl7/resource/PatientMRG.yml:22]: Should we be using the fields in MRG.1 if available?

File: src/main/resources/hl7/resource/PatientMRG.yml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+      display: Medical Record
+      use: old
+
+name:
+   valueOf: datatype/HumanName
+   generateList: true
+   expressionType: resource
+   specs: MRG.7
+
+communication:
```
Comment (LisaWellman) [inline src/main/resources/hl7/resource/PatientMRG.yml:33]: Remove this, we should not be pulling from PID

File: src/main/resources/hl7/resource/PatientMRG.yml [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+   condition: $language NOT_NULL
+   valueOf: secondary/Communication
+   expressionType: resource
+   vars:
+      language: PID.15
+
+active:
+   type: BOOLEAN  
+   valueOf: $isActive
```
Comment (LisaWellman) [inline src/main/resources/hl7/resource/PatientMRG.yml:42]: can you just put 'false' here and delete the constant?

File: src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+        List<Resource> patientResource = e.stream()
+        .filter(v -> ResourceType.Patient== v.getResource().getResourceType())
+        .map(BundleEntryComponent::getResource).collect(Collectors.toList());
+
+        // One for the PID segment and one for the MRG segment
+        assertThat(patientResource).hasSize(2);
+
+        // Get first patient and its id.
+        Resource patientOne = patientResource.get(0);
+        String patientOneId = patientOne.getId();
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java:48]: check for the identifier value "123"

File: src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        
+        // Get second patient and its id.
+        Resource patientTwo = patientResource.get(1);
+        String patientTwoId = patientTwo.getId();
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java:52]: check for the identifier value "456"

Comment (LisaWellman) [reply src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java:52]: check for use=old

File: src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        List<Resource> patientResource = e.stream()
+        .filter(v -> ResourceType.Patient== v.getResource().getResourceType())
+        .map(BundleEntryComponent::getResource).collect(Collectors.toList());
+
+        // One for the PID segment and one for the MRG segment
+        assertThat(patientResource).hasSize(2);
+
+        // Get first patient and its id.
+        Resource patientOne = patientResource.get(0);
+        String patientOneId = patientOne.getId();
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java:118]: check identifier value=MR1

File: src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+        
+        // Get second patient and its id.
+        Resource patientTwo = patientResource.get(1);
+        String patientTwoId = patientTwo.getId();
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java:122]: check identifier value=MR2

Comment (LisaWellman) [reply src/test/java/io/github/linuxforhealth/hl7/segments/HL7MergeFHIRConversionTest.java:122]: check use=old

**Review by LisaWellman — APPROVED** [review]
Approving, you can address some of these comments in the next PR if you prefer.

File: src/main/java/io/github/linuxforhealth/hl7/message/HL7MessageEngine.java (hunk 3 above)
Comment (LisaWellman) [reply src/main/java/io/github/linuxforhealth/hl7/message/HL7MessageEngine.java:98]: Pushpa confirmed she did not mean to commit this and we can revert this change

File: src/main/java/io/github/linuxforhealth/hl7/message/HL7MessageEngine.java [hunk 17]
```diff
@@ -159,10 +175,9 @@ public Bundle transform(final InputDataExtractor dataInput,
   }
 
   private List<ResourceResult> generateResources(HL7MessageData hl7DataInput,
-      HL7FHIRResourceTemplate template, Map<String, EvaluationResult> contextValues,
-      Bundle bundle) {
+      HL7FHIRResourceTemplate template, Map<String, EvaluationResult> contextValues) {
 
-    ResourceModel resourceModel = template.getResource();
+    var resourceModel = template.getResource();
```
Comment (LisaWellman) [inline src/main/java/io/github/linuxforhealth/hl7/message/HL7MessageEngine.java:180]: Please also revert this change

File: src/main/resources/hl7/resource/PatientMRG.yml (hunk 5 above)
Comment (xpander53) [reply src/main/resources/hl7/resource/PatientMRG.yml:2]: Done
